src/ndarray/ndarray.cc

```diff
@@ -1580,13 +1581,20 @@ static const uint32_t NDARRAY_V1_MAGIC = 0xF993fac8;
 /* magic number for ndarray version 2, with storage type */
 static const uint32_t NDARRAY_V2_MAGIC = 0xF993fac9;
 
+// magic number for ndarray version 3, with np shape semantics.
+// The ndarray must be saved and loaded within np shape semantics.
+static const uint32_t NDARRAY_V3_MAGIC = 0xF993faca;
```

Comment:
Do we have any tests for handling legacy storage types?

Reply:
Yes. It's here.
https://github.com/apache/incubator-mxnet/pull/15090/files#diff-69757562d07268150de8b369ff5b6b61R1725